File: web/apps/RoadHelp.Api/Models/StatusChangeLog.cs
using System;
using RoadHelp.Api.Enums;

namespace RoadHelp.Api.Models;

public class StatusChangeLog
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public StatusChangeTargetType TargetType { get; set; }
    public Guid TargetId { get; set; }
    
    public string? OldStatus { get; set; }
    public string? NewStatus { get; set; }
    public string? Reason { get; set; }
    public Guid? ChangedById { get; set; }
    
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

File: web/apps/api/app/models/status_change_log.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Enum, ForeignKey, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base
from app.models.enums import StatusChangeTargetType


class StatusChangeLog(Base):
    __tablename__ = "status_change_logs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    target_type: Mapped[StatusChangeTargetType] = mapped_column(
        Enum(StatusChangeTargetType, name="status_change_target_type", native_enum=True),
        nullable=False,
    )
    target_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), nullable=False, index=True
    )
    old_status: Mapped[str | None] = mapped_column(String(40), nullable=True)
    new_status: Mapped[str] = mapped_column(String(40), nullable=False)
    reason: Mapped[str | None] = mapped_column(Text, nullable=True)
    changed_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        nullable=False, server_default=func.now(), index=True
    )
